sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -85,7 +85,7 @@ object DateTimeUtils {
     // SPARK-6785: use Math.floor so negative number of days (dates before 1970)
     // will correctly work as input for function toJavaDate(Int)
     val millisLocal = millisUtc + timeZone.getOffset(millisUtc)
-    Math.floor(millisLocal.toDouble / MILLIS_PER_DAY).toInt
+    Math.floorDiv(millisLocal, MILLIS_PER_DAY).toInt
```

Comment:
I see, because `floorDiv` rounds towards 0?
What about the comment above, that seems to indicate this might be on purpose?
Generally I agree that `floorDiv` could be appropriate in any place the input could be negative; maybe there are more.

Reply:
> I see, because `floorDiv` rounds towards 0?

For example, if we have an old date (negative in milliseconds) `<old year>-12-31 23:59:59.999456`. `floorDiv` should produce `<old year>-12-31 23:59:59.999` instead of `<old year+1>-01-01 00:00:00.000`.

Reply:
Nevermind, I misunderstood `floorDiv` (I had never seen this method!) I had it backwards; integer division truncates towards 0 when negative, but `floorDiv` doesn't.

The original change here had some tests that in theory were asserting the current behavior. I wonder if they'll fail.

There are also comments in the second change about what the result should be in millis vs micros. Should that be a test case? It doesn't look like it is, and it would help verify the fix.

Reply:
> Should that be a test case?

Added the test case